dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp

```diff
@@ -142,156 +117,158 @@ bool DMFileReader::getSkippedRows(size_t & skip_rows)
+    /// 4. Read and filter packs
 
-    size_t read_rows = 0;
-    for (size_t i = next_pack_id; i < next_pack_id_cp; ++i)
+    MutableColumns columns;
+    columns.reserve(read_columns.size());
+    size_t total_passed_count = countBytesInFilter(filter);
+    for (const auto & cd : read_columns)
+    {
+        auto col = cd.type->createColumn();
+        col->reserve(total_passed_count);
+        columns.emplace_back(std::move(col));
+    }
+
+    size_t offset = 0;
+    // reset next_pack_id to start_pack_id, next_row_offset to start_row_offset
+    next_pack_id = start_pack_id;
+    next_row_offset = start_row_offset;
+    for (size_t pack_id = start_pack_id; pack_id < last_pack_id; ++pack_id)
     {
         // When the next pack is not used or the pack is the last pack, call read() to read theses packs and filter them
         // For example:
         //  When next_pack_id_cp = use_packs.size() and use_packs[next_pack_id:next_pack_id_cp] = [true, true, false, true, true, true]
         //  The algorithm runs as follows:
         //      When i = next_pack_id + 2, call read() to read {next_pack_id, next_pack_id + 1}th packs
         //      When i = next_pack_id + 5, call read() to read {next_pack_id + 3, next_pack_id + 4, next_pack_id + 5}th packs
-        if (use_packs[i] && (i + 1 == use_packs.size() || !use_packs[i + 1]))
+        if (use_packs[pack_id] && (pack_id + 1 == use_packs.size() || !use_packs[pack_id + 1]))
         {
             Block block = read();
+            size_t rows = block.rows();
 
-            IColumn::Filter block_filter(filter.cbegin() + read_rows, filter.cbegin() + read_rows + block.rows());
-            read_rows += block.rows();
-
-            if (size_t passed_count = countBytesInFilter(block_filter); passed_count != block.rows())
+            if (size_t passed_count = countBytesInFilter(filter, offset, rows); passed_count != rows)
             {
-                for (auto & col : block)
+                std::vector<size_t> positions;
+                positions.reserve(passed_count);
+                for (size_t p = offset; p < offset + rows; ++p)
                 {
-                    col.column = col.column->filter(block_filter, passed_count);
+                    if (filter[p])
+                        positions.push_back(p - offset);
+                }
+                for (size_t i = 0; i < block.columns(); ++i)
+                {
+                    columns[i]->insertDisjunctFrom(*block.getByPosition(i).column, positions);
                 }
             }
-
-            blocks.emplace_back(std::move(block));
+            else
+            {
+                for (size_t i = 0; i < block.columns(); ++i)
+                {
+                    columns[i]->insertRangeFrom(
+                        *block.getByPosition(i).column,
+                        0,
+                        block.getByPosition(i).column->size());
+                }
+            }
```

Comment:
We assume `readWithFilter` will never be called on the three hidden columns, so we don't need to take care about whether the `block.getByPosition(i).column` is ColumnConst or not?

Reply:
<img width="871" alt="image" src="https://github.com/pingcap/tiflash/assets/60744015/900c3e6d-582b-4f2c-95c6-c65444ffc057">

clean read is not enabled when there are hidden columns.